File: Android/framework/src/test/java/UnitOfWorkTests.java
import com.DataManipulation.BaseMapperInterfaces.IBaseMapper;
import com.DataManipulation.BaseQueryObjectInterfaces.IBaseQueryObjectConcrete;
import com.DataSynchronizationManager;
import com.Interfaces.IDataSynchronizationManager;
import com.Interfaces.IRepository;
import com.Interfaces.IUnitOfWork;
import com.UnitOfWork;

import java.rmi.NoSuchObjectException;
import java.util.ArrayList;
import java.util.List;

import CustomerServices.Customer;
import CustomerServices.CustomerInvocationDelegates;
import CustomerServices.CustomerMapper;
import CustomerServices.GetCustomerByCivilStatusQuery;
import CustomerServices.GetCustomerByIdQuery;
import CustomerServices.CustomerUoWDelegate;

import static org.junit.Assert.assertEquals;

/**
 * Created by aiko on 4/8/17.
 */

public class UnitOfWorkTests {
    private IDataSynchronizationManager _manager;
    private CustomerMapper _mapper = new CustomerMapper();

    @org.junit.Before
    public void Initialize()
    {
        _manager = DataSynchronizationManager.GetInstance();

        List<IBaseQueryObjectConcrete<Customer>> customerQueryObjects = new ArrayList<IBaseQueryObjectConcrete<Customer>>();

        customerQueryObjects.add(new GetCustomerByIdQuery());
        customerQueryObjects.add(new GetCustomerByCivilStatusQuery());
        _manager.RegisterEntity(Customer.class, new CustomerMapper(), customerQueryObjects);
    }

    @org.junit.Test
    public void TestCommit_SimpleInsertion()
    {
        IRepository<Customer> repository = _manager.GetRepository(Customer.class);
        GetCustomerByCivilStatusQuery.Criteria criteriaByStatus = new GetCustomerByCivilStatusQuery.Criteria(GetCustomerByCivilStatusQuery.CivilStatus.Married);
        GetCustomerByIdQuery.Criteria criteriaById = new GetCustomerByIdQuery.Criteria(2);
        List<Customer> customerResults = new ArrayList<Customer>();
        IUnitOfWork uow = new UnitOfWork();

        customerResults.addAll(repository.Matching(criteriaByStatus));
        customerResults.addAll(repository.Matching(criteriaById));


        IBaseMapper mapper = DataSynchronizationManager.GetInstance().GetMapper(Customer.class);
        Customer customer1 = new Customer(mapper, "1");
        Customer customer2 = new Customer(mapper, "2");
        Customer customer3 = new Customer(mapper, "3");

        CustomerInvocationDelegates delegates = new CustomerInvocationDelegates();

        try
        {
            //Sequence of observation affects commit order
            uow.RegisterNew(customer1, delegates);
            uow.RegisterNew(customer3, delegates);
            uow.RegisterNew(customer2, delegates);
        }
        catch (NoSuchObjectException ex)
        {
        }

        List<String> sequenceDescription = new ArrayList<String>();

        uow.Commit(new CustomerUoWDelegate(sequenceDescription));

        assertEquals("1=Insert=CustomerServices.Customer", sequenceDescription.get(0));
        assertEquals("3=Insert=CustomerServices.Customer", sequenceDescription.get(1));
        assertEquals("2=Insert=CustomerServices.Customer", sequenceDescription.get(2));
    }
}

File: DotNet/CSharp/Framework.Tests/UnitOfWorkTests.cs
﻿using System.Collections.Generic;
using Framework.Data_Manipulation;
using Framework.Tests.CustomerServices;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using static Framework.Tests.CustomerServices.GetCustomersByCivilStatusQuery.Criteria;

namespace Framework.Tests
{
    [TestClass]
    public class UnitOfWorkTests
    {
        private IDomainObjectManager _manager = new DomainObjectManager();
        private CustomerMapper _mapper = new CustomerMapper();

        [TestInitialize]
        public void Initialize()
        {
            _manager.RegisterEntity(
                _mapper,
                new List<IBaseQueryObject<Customer>> {
                    {new GetCustomerByIdQuery()},
                    {new GetCustomersByCivilStatusQuery()}
                });
        }

        [TestMethod]
        public void TestCommit_SimpleInsertion()
        {
            IRepository<Customer> repository = _manager.GetRepository<Customer>();
            GetCustomersByCivilStatusQuery.Criteria criteriaByStatus = GetCustomersByCivilStatusQuery.Criteria.SearchByStatus(CivilStatus.Married);
            GetCustomerByIdQuery.Criteria criteriaById = GetCustomerByIdQuery.Criteria.SearchById(2);
            List<Customer> customerResults = new List<Customer>();
            IUnitOfWork uow = new UnitOfWork();

            customerResults.AddRange(repository.Matching(criteriaByStatus));
            customerResults.Add(repository.Matching(criteriaById));

            
            IBaseMapper mapper = _manager.GetMapper<Customer>();
            Customer customer1 = new Customer(mapper, null) { Number = "1" };
            Customer customer2 = new Customer(mapper, null) { Number = "2" };
            Customer customer3 = new Customer(mapper, null) { Number = "3" };


            //Sequence of observation affects commit order
            uow.RegisterNew(customer1);
            uow.RegisterNew(customer3);
            uow.RegisterNew(customer2);
            
            IList<string> sequenceDescription = new List<string>();

            uow.Commit(
                (domainObject, action, results) =>
                {
                    string description = BaseMapper.GetHashValue<string>(results, CustomerMapper.SUCCESS_DESCRIPTION);

                    sequenceDescription.Add(string.Format("{0}={1}={2}", description, action.ToString(), domainObject.Mapper.GetEntityTypeName()));
                },
                (domainObject, action, results) => { }
                );

            Assert.AreEqual("1=Insert=Customer", sequenceDescription[0]);
            Assert.AreEqual("3=Insert=Customer", sequenceDescription[1]);
            Assert.AreEqual("2=Insert=Customer", sequenceDescription[2]);
        }

        [TestMethod]
        public void TestCommit_RemoveInsertRegistration()
        {
            IRepository<Customer> repository = _manager.GetRepository<Customer>();
            GetCustomersByCivilStatusQuery.Criteria criteriaByStatus = GetCustomersByCivilStatusQuery.Criteria.SearchByStatus(CivilStatus.Married);
            GetCustomerByIdQuery.Criteria criteriaById = GetCustomerByIdQuery.Criteria.SearchById(2);
            List<Customer> customerResults = new List<Customer>();
            IUnitOfWork uow = new UnitOfWork();

            customerResults.AddRange(repository.Matching(criteriaByStatus));
            customerResults.Add(repository.Matching(criteriaById));


            IBaseMapper mapper = _manager.GetMapper<Customer>();
            Customer customer1 = new Customer(mapper, null) { Number = "1" };
            Customer customer2 = new Customer(mapper, null) { Number = "2" };
            Customer customer3 = new Customer(mapper, null) { Number = "3" };


            //Sequence of observation affects commit order
            uow.RegisterNew(customer1);
            uow.RegisterNew(customer3);
            uow.RegisterNew(customer2);

            uow.RemoveExistingRegistration(customer1);
            IList<string> sequenceDescription = new List<string>();

            uow.Commit(
                (domainObject, action, results) =>
                {
                    string description = BaseMapper.GetHashValue<string>(results, CustomerMapper.SUCCESS_DESCRIPTION);

                    sequenceDescription.Add(string.Format("{0}={1}={2}", description, action.ToString(), domainObject.Mapper.GetEntityTypeName()));
                },
                (domainObject, action, results) => { }
                );

            Assert.AreEqual("3=Insert=Customer", sequenceDescription[0]);
            Assert.AreEqual("2=Insert=Customer", sequenceDescription[1]);
        }
    }
}
